### PR TITLE
<functional> tag_invoke

### PR DESCRIPTION
**Pull request by miscco**
This implements tag_invoke and friends

### CODE REVIEW

File: include/p2300/tag_invoke.hpp [hunk 1]
```diff
@@ -28,9 +28,71 @@
...
+namespace std {
+    template <auto& _Tag>
+    using tag_t = decay_t<decltype(_Tag)>;
+    namespace _Tag_invoke {
+        void tag_invoke();
+
+        template <class _Ty>
+        concept _Has_class_or_enum_type = __is_class(remove_reference_t<_Ty>)
+            || __is_enum(remove_reference_t<_Ty>)
+            || __is_union(remove_reference_t<_Ty>);
```
Comment (dietmarkuehl) [inline include/p2300/tag_invoke.hpp]: The names `__is_class`, `__is_enum`, and `__is_union` aren't standard library names and they are not defined here. I did verify that the code compiles with `clang++` and `g++` as is, i.e., they both seem to agree on the same magic name. The code could just use the standard names:

        template <class _Ty>
        concept _Has_class_or_enum_type = is_class_v<remove_reference_t<_Ty>>
            || is_enum_v<remove_reference_t<_Ty>>
            || is_union_v<remove_reference_t<_Ty>>;

Is the reason the special names are used that their use is more efficient when compiling?

File: include/p2300/tag_invoke.hpp [hunk 2]
```diff
@@ -28,9 +28,71 @@
...
+
+        template <class _Tag, class... _Args>
+        concept _Nothrow_tag_invocable_no_adl = _Has_class_or_enum_type<_Tag> && requires {
+            {tag_invoke(std::declval<_Tag>(), std::declval<_Args>()...)} noexcept;
+        };
+
+        template <class _Tag, class _Arg>
+        concept _Tag_observable_no_adl = _Has_class_or_enum_type<_Tag> && requires {
+            {tag_invoke(std::declval<_Tag>(), std::declval<add_const_t<_Arg>>())} noexcept;
+        };
```
Comment (dietmarkuehl) [inline include/p2300/tag_invoke.hpp:57]: The name `_Tag_observable` isn't mention in P2300: is this concept used for the implementation of other components? The defintion/name seems to imply it is used for observer/accessor operations to verify that a `const`-qualifed second argument would also work.

File: test/tag_invoke.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
+// test/call_result_t.cpp                                             -*-C++-*-
```
Comment (dietmarkuehl) [inline test/tag_invoke.cpp]: The files do contain their respective filename to potentially make sense out of context. If we don't want that, I can remove them. If we want to retain that, I think the name should be adjusted.

**Review by dietmarkuehl — COMMENTED** [review]
It turns out, I had implemented `tag_invoke` but not, yet, pushed as I hadn't completed tests. I'm happy to accept this version and there are just two comments which are more questions.

One issue is the copyright: I didn't really write this code and acquiring the copyright for it seems wrong. I'm not sure how we want to deal with that? Not just for this file but in general. Maybe it would be reasonable to let the Standard C++ Foundation be the copyright owner? I wouldn't change that without asking them first, though (and I think whoever is the copyright owner can just change the statement).

File: include/p2300/tag_invoke.hpp (hunk 1 above)
Comment (miscco) [reply include/p2300/tag_invoke.hpp]: I think in this case it is totally fine.

Sometimes the `is_meow_v` traits are defined as `is_meow::value` where it indeed has a throughput benefit. I can use the proper traits

File: include/p2300/tag_invoke.hpp (hunk 2 above)
Comment (miscco) [reply include/p2300/tag_invoke.hpp:57]: Yeah, I should have maybe named it `_Tag_querriable` but that is a horrible word.

In a nutshell it should cover all querries that take the CPO by const reference and are noexcept. We have a few of them and centralizing them reduces the risk of errors

File: test/tag_invoke.cpp (hunk 3 above)
Comment (miscco) [reply test/tag_invoke.cpp]: Copy & paste error here

File: include/p2300/tag_invoke.hpp (hunk 2 above)
Comment (dietmarkuehl) [reply include/p2300/tag_invoke.hpp:57]: OK. I wasn't worried about the [internal] name. I just wondered why this is here as the spec doesn't say anything about it and I didn't see it used [yet].

**Review by dietmarkuehl — APPROVED** [review]
Looks good to me.

**Pull request merged**